lib/PipelineC/Tracing/Runner.cpp

```diff
@@ -0,0 +1,402 @@
+    llvm::SmallString<128> Output;
+    llvm::sys::fs::createUniqueDirectory("/tmp/tmp.revng-run-trace", Output);
+    TempDirs.push_back(Output.str().str());
+    return Output.str().str();
+  }
+
+  template<typename T>
+  void storePointer(const llvm::StringRef Name, T *Pointer) {
+    revng_assert(std::is_pointer_v<T>);
+    if (Name == "P0x0") {
+      return;
+    }
+    Pointers[Name.str()] = reinterpret_cast<uintptr_t>(Pointer);
+  }
+
+  template<typename T>
+  T getPointer(const llvm::StringRef Name, bool Invalidate) {
+    revng_assert(std::is_pointer_v<T>);
+    revng_assert(Pointers.count(Name));
+
+    T Result = (T) Pointers[Name];
+    if (Invalidate)
+      Pointers.erase(Name);
+
+    return Result;
+  }
+
+  void reset() {
+    for (auto &Dir : TempDirs) {
+      std::filesystem::remove_all(Dir);
+    }
+    TempDirs.clear();
+    Pointers.clear();
+  }
+};
+
+using ArgumentsRef = const llvm::ArrayRef<CommandArgument>;
+using ReturnRef = const llvm::StringRef;
+
+// clang-format off
+template<typename T>
+using PlainStorage = std::conditional_t<anyOf<T, char *, const char *>(),
+                                        std::vector<char>,
+                                        T>;
+template<typename T>
+using ArgStorage = std::conditional_t<isList<T>(),
+                                      std::vector<std::remove_pointer_t<T>>,
+                                      PlainStorage<T>>;
+
+using CommandRunner = std::function<void(RunnerContext &Context,
+                                         ArgumentsRef Arguments,
+                                         ReturnRef Return)>;
+// clang-format on
+
+static std::map<std::string, CommandRunner> CommandHandler;
+
+namespace utils {
+// Utility functions for decoding arguments
+
+template<typename T>
+inline T toInt(const llvm::StringRef StrInt) {
+  T Result;
+  StrInt.getAsInteger(10, Result);
+  return Result;
+}
+
+template<typename T>
+inline T toInt(ArgumentsRef Args, const size_t Idx) {
+  revng_assert(Args[Idx].isScalar());
+  return toInt<T>(Args[Idx].Scalar);
+}
+
+inline std::vector<const char *>
+toStringList(ArgumentsRef Args, const size_t Idx) {
+  revng_assert(Args[Idx].isSequence());
+  std::vector<const char *> Ret;
+  for (auto &Elem : Args[Idx].Sequence) {
+    Ret.push_back(Elem.c_str());
+  }
+  // Used to catch OOB errors earlier
+  Ret.push_back(nullptr);
+  return Ret;
+}
+
+template<typename T>
+inline std::vector<T> toIntList(ArgumentsRef Args, const size_t Idx) {
+  revng_assert(Args[Idx].isSequence());
+  std::vector<T> Ret;
+  for (auto &Elem : Args[Idx].Sequence) {
+    Ret.push_back(toInt<T>(Elem));
+  }
+  return Ret;
+}
+
+inline std::vector<std::string>
+parseRelativePaths(ArgumentsRef Args, const size_t Idx) {
+  revng_assert(Args[Idx].isSequence());
+  std::vector<std::string> Result;
+  for (auto &Elem : Args[Idx].Sequence) {
+    llvm::StringRef Element(Elem);
+    auto [Prefix, Path] = Element.rsplit("/root/");
```

Comment:
Eh, but then we're making the assumption that the paths in functions calls is always relative
I was thinking of putting a `metdata` section in the trace where the trace puts `getCurrentRoot`, then we just check if paths are a subpath and in case we replace

Reply:
It's also for pipeline `.yml` files...
I understand the need, but I'd keep the `trace run` as low-friction as possible

Reply:
One place or another we need to correct it. We are already overriding `rp_initialize` and `rp_manager_create` to do ad-hoc things for paths.
Just blanket replacing those two lists seems a simpler solution than introducing `getCurrentRoot` in the YAML file and patching it in place... It brings no additional benefit.

Reply:
Main considerations against blanket replacing:
* `collect_files` is done in python, we'd have to wrap it or rewrite it in c++
* `revng trace run` cannot use llvm's command line parsing, which makes adding a bunch of `-load` kind of annoying
* Until the trace is loaded there's no way of saying things like "Only revng pipelines/`.so`s were loaded, which might introduce inconsistencies

All of these are not deal-breakes, but the path of least resistance (to me at least) seems to just record the root when tracing and the prefix-replace the entries when running

Reply:
As per our discussion, let's drop from `rp_initialize` and `rp_manager_create` paths altogether and let's use `cl::opt` in `revng-trace-run` too.

Reply:
Dome